internals/secrethub/read_test.go

```diff
@@ -1,9 +1,122 @@
+	"github.com/secrethub/secrethub-go/internals/errio"
+	"github.com/secrethub/secrethub-go/pkg/secrethub"
+	"github.com/secrethub/secrethub-go/pkg/secrethub/fakeclient"
 )
 
 func TestReadCommand_Run(t *testing.T) {
-	// TODO SHDEV-1029 Test ReadCommand.
+	testErr := errio.Namespace("test").Code("test").Error("test error")
+	testSecret := []byte("Remember! Reality's an illusion, the universe is a hologram, buy gold! Bye! - Bill Cipher")
+
+	cases := map[string]struct {
+		cmd             ReadCommand
+		newClientErr    error
+		secretVersion   api.SecretVersion
+		serviceErr      error
+		expectedContent string
+		out             string
+		err             error
+	}{
+		"success read": {
+			cmd: ReadCommand{
+				path: "test/repo/secret",
+			},
+			secretVersion: api.SecretVersion{Data: testSecret},
+			out:           string(testSecret) + "\n",
+		},
+		"success clipboard": {
+			cmd: ReadCommand{
+				path:                "test/repo/secret",
+				clipper:             fakeclip.New(),
+				useClipboard:        true,
+				clearClipboardAfter: 5 * time.Minute,
+			},
+			secretVersion:   api.SecretVersion{Data: testSecret},
+			expectedContent: string(testSecret),
+			out:             "Copied test/repo/secret to clipboard. It will be cleared after 5 minutes.\n",
+		},
+		"success file": {
+			cmd: ReadCommand{
+				path:     "test/repo/secret",
+				outFile:  "secret.txt",
+				fileMode: filemode.New(os.ModePerm),
+			},
+			secretVersion:   api.SecretVersion{Data: testSecret},
+			expectedContent: string(testSecret) + "\n",
+			out:             "",
+		},
+		"fail file": {
+			cmd: ReadCommand{
+				path:     "test/repo/secret",
+				outFile:  "/fail/read.txt",
+				fileMode: filemode.New(os.ModeAppend),
+			},
+			secretVersion: api.SecretVersion{Data: testSecret},
+			out:           "",
+			err:           ErrCannotWrite("/fail/read.txt", "open /fail/read.txt: no such file or directory"),
+		},
+		"new client error": {
+			newClientErr: testErr,
+			err:          testErr,
+		},
+		"read error": {
+			serviceErr: testErr,
+			err:        testErr,
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			// Setup
+			testIO := fakeui.NewIO(t)
+			tc.cmd.io = testIO
+
+			if tc.newClientErr != nil {
+				tc.cmd.newClient = func() (secrethub.ClientInterface, error) {
+					return nil, tc.newClientErr
+				}
+			} else {
+				tc.cmd.newClient = func() (secrethub.ClientInterface, error) {
+					return fakeclient.Client{
+						SecretService: &fakeclient.SecretService{
+							VersionService: &fakeclient.SecretVersionService{
+								GetWithDataFunc: func(path string) (*api.SecretVersion, error) {
+									return &tc.secretVersion, tc.serviceErr
+								},
+							},
+						},
+					}, nil
+				}
+			}
+
+			// Run
+			err := tc.cmd.Run()
+			content := ""
+			if _, err := os.Stat(tc.cmd.outFile); err == nil {
+				res, _ := ioutil.ReadFile(tc.cmd.outFile)
+				content = string(res)
+				os.Remove(tc.cmd.outFile)
```

Comment:
It is not recommended to use files in test cases for several reasons:

1. You might not clean up the file correctly if the test fails. In this case what would happen if `os.Stat` or `ioutil.ReadFile` returned an error? In either cases the out file wouldn't be deleted.

2. Your test cases are no longer isolated. Since all test cases would access the same file system, this would effectively behave like shared state and could bring about errors that are really hard to find and might even depend on the order of execution of the test cases.

3. It is generally good to have as much control over the outcome of the test as possible. In this case the calls to the os might sometimes fail for various reasons (no disk space left, a file with the same name already exists etc.). We only want the test to fail if our code is incorrect.

So to avoid this the best solution would be to replace the `ioutil.WriteFile` call in the read command's `Run` method with another function while testing, just like we're replacing `client.New` with other functions by setting the `newClientFunc` variable.

Note that in other situations, if writing to the file system is absolutely necessary, temporary files can be used when testing as these circumvent most of the issues listed here.

Reply:
👍  on Marton's comment. Free tip: one way to use files in tests (when it is necessary) is by leveraging [ioutil.TempDir](https://golang.org/pkg/io/ioutil/#ReadDir)

Reply:
Done